\documentclass{bioinfo}

\usepackage{graphicx}
\usepackage{color} % Only required for the TODO macro.
\usepackage{amssymb} % Symbols like checkmark

%\usepackage{hyperref} % links, but no colored boxes
%\usepackage{breakurl}
\usepackage[hyphens]{url}
%\hypersetup{breaklinks=true,
%pagecolor=white,
%colorlinks=false}
\urlstyle{rm} %so it doesn't use a typewriter font for url’s.

% Allow line breaks in texttt
\newcommand{\origttfamily}{}
\let\origttfamily=\ttfamily
\renewcommand{\ttfamily}{\origttfamily \hyphenchar\font=`\-}

\hyphenation{BioPAX BioCarta PaxTools SBML net-work net-works er-ro-neous-ly
      Clandestine re-la-tigon re-la-tigon-ship re-ac-tigon other-wise
      re-la-tins re-la-tigon-ships re-ac-tins inter-action inter-actions re-search
      be-tween}

\newcommand{\NEW}[1]{\textcolor{red}{#1}}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Useful macros:

%\newcommand{\TTra}{\textsuperscript{\tiny{}}}
\newcommand{\TODO}[1]{\textcolor{red}{\textbf{#1}}}

% B
\newcommand{\BiochemicalReaction}{\texttt{Bio\-chemical\-Reaction}}

% C
\newcommand{\Catalysis}{\texttt{Cata\-lysis}}
\newcommand{\ComplexAssembly}{\texttt{Complex\-Assembly}}
\newcommand{\Conversion}{\texttt{Conversion}}
\newcommand{\Control}{\texttt{Control}}
\newcommand{\Controller}{\texttt{Controller}}
\newcommand{\Controlled}{\texttt{Controlled}}
\newcommand{\ControlType}{\texttt{Control\-Type}}

% D
\newcommand{\Degradation}{\texttt{Degra\-dation}}
\newcommand{\DNAregion}{\texttt{DNAregion}}

% E
\newcommand{\Entity}{\texttt{Entity}}

% F
\newcommand{\functionTerm}{\texttt{function\-Term}}

% G
\newcommand{\GeneticInteraction}{\texttt{Genetic\-Interaction}}

% I
\newcommand{\inputP}{\texttt{input}}
\newcommand{\Interaction}{\texttt{Inter\-action}}

% M
\newcommand{\model}{\texttt{model}}
\newcommand{\Modulation}{\texttt{Modulation}}
\newcommand{\MolecularInteraction}{\texttt{Mo\-le\-cu\-lar\-In\-ter\-ac\-tion}}

% O
\newcommand{\outputP}{\texttt{output}}

% P
\newcommand{\Pathway}{\texttt{Pathway}}
\newcommand{\PhysicalEntity}{\texttt{Physical\-Entity}}

% Q
\newcommand{\qual}{qual}
\newcommand{\qualitativeModel}{\texttt{qualitative\-Model}}
\newcommand{\qualitativeSpecies}{\texttt{qualitative\-Species}}

% R
\newcommand{\reaction}{\texttt{reaction}}
\newcommand{\reactions}{\texttt{reactions}}
\newcommand{\RNAregion}{\texttt{RNAregion}}

% S
\newcommand{\sign}{\texttt{sign}}
\newcommand{\species}{\texttt{species}}
\newcommand{\Symbol}{\texttt{symbol}}

% T
\newcommand{\TemplateReactionRegulation}{\texttt{Template\-Reaction\-Regulation}}
\newcommand{\TemplateReaction}{\texttt{Template\-Reaction}}
\newcommand{\transition}{\texttt{transition}}
\newcommand{\transitions}{\texttt{transitions}}
\newcommand{\Transport}{\texttt{Transport}}
\newcommand{\TransportWithBiochemicalReaction}{\texttt{Transport\-With\-Biochemical\-Reaction}}

% End macros.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\copyrightyear{2012}
\pubyear{2012}

\begin{document}

\firstpage{1}

\title[BioPAX to SBML qual]{Qualitative translation of relations from BioPAX to SBML qual}
\author[Finja B\"uchel \textit{et~al.}]{
Finja B\"uchel\,$^{1\,,}$\footnote{to whom correspondence should be
addressed}\,, Clemens Wrzodek\,$^1$,
Florian Mittag\,$^1$,
Andreas Dr\"ager\,$^1$,
Johannes Eichner\,$^1$,
Nicolas Rodriguez\,$^2$,
Nicolas Le Nov\`{e}re\,$^2$,
and Andreas~Zell\,$^1$}
\address{$^{1}$Center for Bioinformatics Tuebingen (ZBIT), University of Tuebingen, T\"ubingen, Germany
\\
$^{2}$Computational Systems Neurobiology Group, European Bioinformatics Institute, Hinxton, United~Kingdom
}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\maketitle

\begin{abstract}
\section{Motivation:}
The Biological Pathway Exchange Language (BioPAX) and the Systems Biology Markup Language (SBML) belong to the most popular modeling and data exchange languages in systems biology.
The focus of SBML is quantitative modeling and dynamic simulation of models, whereas the BioPAX specification concentrates mainly on visualization and qualitative analysis of pathway maps.
BioPAX describes reactions and relations.
In contrast, reactions are the only type of interaction in core SBML.
With the SBML Qualitative Models extension (\qual), it has recently also become possible to describe relations in SBML.
Before the creation of \qual{}, relations could not be \NEW{properly translated into SBML or just by adding additional SBO terms on reactions which is not so proper}.
%Before the creation of \qual{}, relations could not be translated into SBML at all or were erroneously converted into reactions}.
Until now, there exists no BioPAX to SBML converter that is fully capable of translating both reactions and relations.
\section{Results:}
The entire Nature Pathway Interaction Database (PID), which includes pathways from BioCarta, Reactome, and the National Cancer Institute, has been converted from BioPAX into SBML.
All available PID BioPAX pathway files (Level~2 and Level~3) have been translated into the SBML format (Level~3 Version~1) including both reactions and relations by using the new \qual{} extension package.
Additionally, we present the new webtool BioPAX2SBML, which can be used for further BioPAX to SBML conversions.
Compared to previous conversion tools, BioPAX2SBML is more comprehensive, more robust and more exact.
\section{Availability:}
BioPAX2SBML is freely available at \url{http://webservices.cs.uni-tuebingen.de/} and the complete collection of the PID models at \url{http://www.cogsys.cs.uni-tuebingen.de/downloads/Qualitative-Models/}.
\section{Contact:}
\href{mailto:finja.buechel@uni-tuebingen.de}{finja.buechel@uni-tuebingen.de}
\NEW{\section{Supplementary Information:}
Supplementary data are available at Bioinformatics online. }
\end{abstract}




\section{Introduction}
The goal of systems biology is the model-driven understanding of biological and biochemical processes across all layers and various levels of detail.
The Biological Pathway Exchange Language (BioPAX) and the Systems Biology Markup Language (SBML) are common modeling languages that facilitate the exchange and storage of \emph{in-silico} models.
BioPAX can be used to describe the biological \NEW{semantics} of metabolic, signaling, molecular, gene-regulatory and genetic interaction networks \citep{Demir2010}.
SBML describes the structure of models, and offers the possibility to include mathematical expressions \citep{Hucka2003}.
The SBML core specification defines reactions in detail but no other relationships between molecules.
Those relationships are denoted as \emph{relations} that specify enzyme-enzyme relations, protein-protein interactions, interactions of transcription factors and genes, protein-compound interaction, links to other pathways, etc.
Before the creation of the Qualitative Models extension for SBML (\qual, \citealp[see][]{QualSpecification}), it was not possible to define relations or to include reactions together with relations in one model.
Furthermore, the combination or exchange of information between different databases is hardly feasible if one database uses the BioPAX format and the other one the SBML format.
So far, there exist several visualization tools, such as Cytoscape \citep{Zinovyev2008, Smoot2011a} or Clandestine \citep{Funahashi2007, Mi2011}, which can handle BioPAX and SBML by using plugins. But the combination of both formats or conversion of one format into the other is difficult even with these plugins.
Hence, converters are needed.

Today, there exist mainly converters from SBML to BioPAX like \emph{The System Biology Format Converter} \citep[see][]{SBFC}, but no converter for BioPAX to SBML that is capable of properly including relations.
Other research groups previously faced the same problem with incompatibilities between BioPAX and SBML.
To overcome the limitations of those file formats and to avoid the creation of pseudo-reactions or similar constructions, \citet{Ruebenacker2009} introduce an intermediate bridging format.
%
The need to combine both formats to use the knowledge from a multitude of databases in various applications becomes more and more urgent.

In this paper we present a webtool for the translation from BioPAX into SBML format. We demonstrate its functionality by converting the whole Nature Pathway Interaction Database (PID, \citealp{Schaefer2009}) from BioPAX Level~2 and Level~3 formats to the SBML format, including the \qual{} extension.


\begin{methods}
\section{Material and Methods}
\subsection{SBML and the Qualitative Models extension}
The SBML Level~3 Version~1 core specification defines a special XML dialect to describe quantitative models.
The most important classes are \species{}, describing reactive species, and \reactions{}, which interconnect \species{} elements.
A \species{} element can be further specified with the aid of MIRIAM annotations \citep{Novere2005}.
%The SBML core specification provides no possibility to define other relationships than concrete quantitative reactions.
\NEW{The SBML core specification provides several constructs to describe quantitative processes, suchs as events, rules, constraints, reactions, etc., but there is no possibility to define qualitative relationships.}

The SBML Qualitative Models extension (\qual) introduces qualitative elements, such as \qualitativeSpecies{} and \transition, providing the necessary means to describe relationships that \NEW{cannot} be described by reactions, for instance, enzyme-enzyme relations or interactions of transcription factors and genes \citep{QualSpecification}.
Instead of the quantities associated to \species, which are affected by reactions, \qualitativeSpecies{} exhibit discrete states, representing their activities that are changed using transitions.
These transitions are linked to \inputP{} and \outputP{} elements.
The \sign{} attribute of the \inputP{} elements describes whether the relationship between the \inputP{} and \outputP{} elements is \emph{positive}, \emph{negative}, \emph{dual}, or \emph{unknown}.
\emph{Dual} means that the transition can operate both activating (\emph{positive}) and inhibiting (\emph{negative}).
In contrast, \emph{unknown} is assigned to the \inputP{} if the transition effect is not further specified.
If, in a qualitative model, the activity of protein A inhibits the activity of protein B, this would be represented as a transition with an input A, whose \sign{} attribute is \emph{negative}, and an output B.


\subsection{The BioPAX specification}
The Biology Pathway Exchange Language (BioPAX) is an OWL (Web Ontology Language) dialect based on RDF.
There is one superclass called \Entity{} that is extended by all other BioPAX classes.
Two main classes are distinguished: \PhysicalEntity{} and \Interaction.
\PhysicalEntity{} describes molecules, such as proteins, complexes, small molecules, DNA, or RNA, whereas \Interaction{} defines reactions and relations between \PhysicalEntity{} classes.
%\Interaction{} is split into \Control{} and \Conversion{}, which can be separated into the subclasses \Catalysis, \Modulation, \TemplateReactionRegulation, \TemplateReaction, \GeneticInteraction, \MolecularInteraction, \Transport, \BiochemicalReaction, \TransportWithBiochemicalReaction, \Degradation{}, and \ComplexAssembly.
\Interaction{} is split into \Control{} and \Conversion{}, which can be separated in \NEW{several} subclasses \NEW{(see Figure \ref{fig:BioPAXSBMLqual})}.

BioPAX is released level-wise with the current level being Level~3.
Level~1 is exclusively able to describe metabolic interactions, whereas Level~2 supports signaling pathways and molecular interactions.
In addition to Level~2, gene-regulatory networks and genetic interactions can be described with Level~3.
%For this purpose the \Control{} subclass \TemplateReactionRegulation, the \Conversion{} subclass \Degradation{}, \TemplateReaction, \GeneticInteraction, and \MolecularInteraction{} have been added.
%Furthermore, \PhysicalEntity{} is now able to define \DNAregion{} and \RNAregion.
For this purpose \NEW{several new BioPAX instances} have been added \NEW{see dashed elements in Figure \ref{fig:BioPAXSBMLqual}}.
Level~3 is not downwards compatible with Level~2, but Level~2 is downwards compatible with Level~1 \citep{Demir2010}.
The BioPAX specification of Level~2 denotes all classes in lower case typewriter font and the specification of Level~3 denotes them in upper case typewriter font. For better readability of this paper, all BioPAX element names begin with capital letters and refer to Level~2 and 3. In contrast, SBML classes are written in lower case.




\subsection{Conversion of BioPAX to SBML \qual}\label{subsec:BioPAX}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%% FIGURE 1 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{figure*}[t!h]
\centering \includegraphics[width=0.9\textwidth]{Buechel_Figure1_BioPaxSBMLqual.png}
\caption{Conversion from BioPAX Level~2 and Level~3 to SBML Level~3 Version~1 with the Qualitative Models extension (\qual).
The green rounded rectangles on the righthand site describe the SBML and \qual{} classes, and the blue ones left the BioPAX elements.
The distinction between BioPAX Level~2 and Level~3 elements is visualized with dashed rectangles.
The dashed rectangles denote elements, which are only available in Level~3.
All other elements occur in both levels.
The ancestry of both BioPAX and SBML elements is indicated with arrows.
Lines, ending with a diamond, indicate elements that are contained in other elements.
The conversion from BioPAX to SBML \qual{} is drawn with black lines.
For some BioPAX elements, it depends on the enclosed entities if the BioPAX element is translated into a reaction or to a relation.
This translation dependency is visualized with black dashed lines.
A detailed translation description of those elements is shown in Table~\ref{Tab:BioPAX2SBML}.}
\label{fig:BioPAXSBMLqual}
\end{figure*}
%%%%%%%%%%%%%%%%%%%%%%%%%%% END FIGURE 1 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
The complete Nature Pathway Interaction Database (PID) has been converted from BioPAX to SBML Level~3 Version~1 including the Qualitative Models extension (\qual).
PID provides curated pathways from the National Cancer Institute (L2/L3 2012-03-16), pathways from BioCarta (L2 2009-09-09, L3 2010-08-10), and human Reactome pathways (L2/L3 2010-08-10 \citealt{Schaefer2009}).
The translation of the BioPAX Level~2 and Level~3 pathway files is performed in four steps: (1) initializing the models, (2) translation of \PhysicalEntity{} elements, (3) translation of \Interaction{} elements, and (4) annotation of all \species.
An overview of the mapping from BioPAX elements to SBML and to SBML \qual{} elements is shown in Figure~\ref{fig:BioPAXSBMLqual}.

%%%%%%%%%%%%%% 1. Step %%%%%% determine organism
\subsubsection{Step 1: Initializing the models.}
Firstly, the pathway organism is determined by searching for the \texttt{BioSource} reference in the BioPAX file.
Furthermore, the SBML \model{} and \qualitativeModel{} are built.
Both models correspond to the complete pathway represented in the BioPAX file.

%%%%%%%%%%%%%% 2. Step %%%%%% entities and annotation
\subsubsection{Step 2: Translation of \PhysicalEntity{} elements.}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%% TABLE X %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{table}[tb]
\processtable{BioPAX \Entity{}'s and assigned SBO terms\label{Tab:BioPAX2SBO}}
{\begin{tabular}{llll}\toprule
\textbf{BioPAX \Entity} & \textbf{Assigned SBO term} & \textbf{SBO name} \\
\midrule
Gene            & SBO:0000354 & informational molecule segment \\
Complex         & SBO:0000253 & non-covalent complex \\
Protein         & SBO:0000252 & polypeptide chain \\
Dna             & SBO:0000251 & deoxyribonucleic acid \\
DnaRegion       & SBO:0000251 & deoxyribonucleic acid \\
Rna             & SBO:0000250 & ribonucleic acid \\
RnaRegion       & SBO:0000250 & ribonucleic acid \\
SmallMolecule   & SBO:0000247 & simple chemical \\
\botrule
\end{tabular}}
{
Each BioPAX \Entity{} is converted to an SBML \species{} and \qualitativeSpecies.
In BioPAX, one can specify the nature of the real entity by classes that are derived from \Entity{} (e.g., DNA, Protein, etc).
SBML does not contain specific entities that can be derived from an SBML \species.
The common way to separate different genomic entities in SBML is using SBO terms from the material entity branch.
This table specifies the SBO terms that we used to distinguish between various cellular entities in SBML.
}
\end{table}
%%%%%%%%%%%%%%%%%%%%%%%%%%% END Table 2 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

In this step, an SBML \species{} and \qualitativeSpecies{} are created for each \PhysicalEntity{}.
Depending on the kind of the \PhysicalEntity, i.e., if it is a protein, complex, DNA, RNA, or small molecule, the \species{} is annotated with the corresponding SBO term (\citealp{SBO}).
The used SBO terms are listed in Table~\ref{Tab:BioPAX2SBO}.
Furthermore, the \species{} compartment is assigned due to the \texttt{CellularLocation} of the \PhysicalEntity.
The default compartment is set if the \texttt{CellularLocation} is not known.

Then, the BioPAX document is mined for an RDF link from the \PhysicalEntity{} to a corresponding Entrez Gene ID.
These identifers are unique and facilitate the automated annotation of this \species{} (described in the fourth step).
If there exists no Gene ID but a gene symbol, the gene symbol is mapped to a Gene ID.

%%%%%%%%%%%%%% 3. Step %%%%%% reaction relations
\subsubsection{Step 3: Translation of \Interaction{} elements.}
BioPAX \Interaction{} elements are translated into SBML core \reactions{} and qual \transitions. An SBML \transition{} describes relationships between molecules that \NEW{cannot} be translated into \reactions{}. Examples for such relationships are enzyme-enzyme relations, protein-protein interactions, interactions of transcription factors and genes, protein-compound interaction, links to other pathways, etc.
BioPAX \Interaction{} elements can be split into \Conversion{} and \Control{} elements.

The translation of the \Conversion{} elements is straightforward, because all elements can unambiguously be mapped to SBML \reactions{}.
The translation of those elements is performed by creating the same reaction with all substrates, products and enzymes in SBML.
Furthermore, the stoichiometry of the reactants and products of {\BiochemicalReaction} and {\TransportWithBiochemicalReaction} are also translated into SBML.

%%%%% Control
The translation of \Control{} elements is more %sophisticated
\NEW{complicated}, because they are translated into a \transition{} or a \reaction{} depending on enclosed \Control{} elements.
\Control{} elements always consist of zero or more \Controller{} and zero or one \Controlled{} elements.
\Controller{} elements are inherited from \PhysicalEntity{} or \Pathway, whereas \Controlled{} elements are also \Interaction{} elements.
Thus, it depends on the kind of \Controller{} and \Controlled{} element whether the \Interaction{} is translated into an SBML \reaction{} or \transition.
%If the \Controller{} or the \Controlled{} element is a \Pathway{} element, the \Interaction{} is always converted to a \transition, because it is biologically not possible to create a reaction with a whole pathway as a reactant or product.
%A \Conversion{} is translated into a \transition{} if the \Controlled{} element is translated into a \transition, too.
%For instance, the conversion of a \Modulation, consisting of a \PhysicalEntity{} as \Controller{} and a \BiochemicalReaction{} as \Controlled, is translated into a reaction.
%An example is shown in Figure~\ref{fig:Ceramide}, which shows the ceramide signaling pathway, where the biochemical reaction from sphingomyelin to ceramide (the \Controlled{} element) is positively modulated from SMPD1+ (the \Controller).
%This modulation will be converted into a \reaction{} where SMPD1+ is modeled as an enzyme of the reaction.
%But if the \Controlled{} element is a \GeneticInteraction, the \Modulation{} is converted into a transition.
%A detailed overview of the conversion of the \Control{} elements is shown in Table~\ref{Tab:BioPAX2SBML}.
%The \sign{} attribute of the \inputP{} element describes the relationship between \inputP{} and \outputP{} and is determined depending on the \ControlType{} attribute.
%This attribute is assigned to nearly all \Control{} elements.
%If the \ControlType{} is activating, \sign{} is set to \emph{active}; if it is inhibiting \sign{} is set to \emph{negative}; if it is both, \sign{} is set to \emph{dual}; otherwise \sign{} is set to \emph{unknown}.
\NEW{
All \Controller-\Controlled{} combinations and the corresponding SBML classes are listed in Table~\ref{Tab:BioPAX2SBML} and discussed in more detail in the Supplement.
For nearly all \Control{} elements a \ControlType{} is assigned describing the relationship between the enclosed elements (i.e. activating, inhibiting).
Depending on this type, the \sign{} attribute of the SBML \inputP{} element is determined.
}


%%%%%%%%%%%%%%%% Step 4 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection{Step 4: Annotation of the translated model.}

Finally, the SBML instances are further annotated.
The BioPAX specification allows users to encode arbitrary identifiers for elements.
These can be identifiers for various databases, e.g., UniProt, Entrez Gene, Ensembl, etc.
Unfortunately, the syntax used in BioPAX is sometimes inconsistent, which leads to XML database annotations like ``UniProt'' or ``UniProtKB'' within BioPAX documents that hamper the automatic reading and interpretation of those models by third party applications.

In SBML, such identifiers can be expressed as standardized MIRIAM URNs that can be added as annotation to any SBML element.
We support and add MIRIAM identifiers for the following databases: Entrez Gene, Omim, Ensembl, UniProt, ChEBI, DrugBank, Gene Ontology, HGNC, PubChem, 3DMET, NCBI Taxonomy, PDBeChem, GlycomeDB, LipidBank, EC-Numbers (enzyme nomenclature) and various KEGG databases (gene, glycan, reaction, compound, drug, pathway, orthology).
To obtain identifiers for those databases, we map the Entrez Gene identifier, which we annotated on every element in Step 2, to a KEGG identifier.
Using the KEGG API, we then query all of those identifiers to retrieve more descriptive names, descriptions of the elements, and the mentioned database identifiers.
The goal of those annotations is to provide models whose components can uniquely be identified by any application and be linked to external data sources.
\end{methods}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%% TABLE 1 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{table}[t!h]
\processtable{Description of the translation of BioPAX \Control{}
elements\label{Tab:BioPAX2SBML}} {\begin{tabular}{llll}\toprule
BioPAX            & BioPAX \Controlled              & Converted\\
\Controller       &                                 & SBML \qual\\
                  &                                 & element\\
\midrule
%\textbf{BioPAX Level~3}\\
\multicolumn{3}{c}{\centering \textbf{BioPAX Level~3}}\\
\midrule
PhysicalEntity & BiochemicalReaction                & reaction\\
PhysicalEntity & ComplexAssembly                    & reaction\\
PhysicalEntity & Conversion                         & transition\\
PhysicalEntity & Degradation                        & reaction\\
PhysicalEntity & Transport                          & reaction\\
PhysicalEntity & TransportWithBiochemicalReaction   & reaction\\
PhysicalEntity & Pathway                            & transition\\
PhysicalEntity & TemplateReaction                   & transition\\
\\
Pathway         & BiochemicalReaction               & transition\\
Pathway         & ComplexAssembly                   & transition\\
Pathway         & Conversion                        & transition\\
Pathway         & Degradation                       & transition\\
Pathway         & Pathway                           & transition\\
Pathway         & TemplateReaction                  & transition\\
Pathway         & Transport                         & transition\\
Pathway         & TransportWithBiochemicalReaction  & transition\\
\\\midrule
%\textbf{BioPAX Level~2}\\
\multicolumn{3}{c}{\centering \textbf{BioPAX Level~2}}\\
\midrule
physicalEntity & biochemicalReaction                & reaction\\
physicalEntity & complexAssembly                    & reaction\\
physicalEntity & interaction                        & transition\\
physicalEntity & pathway                            & transition\\
physicalEntity & transport                          & reaction\\
physicalEntity & transportWithBiochemicalReaction   & reaction\\
\\
pathway         & biochemicalReaction               & transition\\
pathway         & complexAssembly                   & transition\\
pathway         & interaction                       & transition\\
pathway         & pathway                           & transition\\
pathway         & transportWithBiochemicalReaction  & transition\\
pathway         & transport                         & transition\\\botrule
\end{tabular}}{BioPAX \Control{} elements consist of a \Controller{} and one
or more \Controlled{} elements.
Depending on the kind of \Controller{} or \Controlled{} element, the \Control{} entity is translated into an SBML reaction or transition.
The table gives an overview of this conversion regarding BioPAX Level~2 and BioPAX Level~3.
}
\end{table}
%%%%%%%%%%%%%%%%%%%%%%%%%%% END TABLE 1 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%



\subsection{Implementation}
The conversion was implemented in Java\texttrademark, using JSBML \citep{Draeger2011} with the Qualitative Models extension, PaxTools \citep{Demir2010}, and the KEGG API \citep{Kanehisa2006}.
PaxTools was used to read the BioPAX files and to manipulate the information content.
With the aid of the KEGG API, this information was extended with MIRIAM identifiers \citep{Novere2005} from the various databases, for instance Entrez Gene, Ensembl, UniProt, etc.



\section{Results and Discussion}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% comparison table %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\newcolumntype{C}{>{\centering\arraybackslash}p{2cm}}
\begin{table*}[htb]
\caption{Comparison of different available converters for BioPAX pathways.}
\label{tab:Comparison}
%\rowcolors{1}{white}{tableShade2}
\makebox[\textwidth]{
\begin{tabular}{lccc}
\toprule
{\bf }                              & {\bf BioPAX2SBML}                 & {\bf Sybill}                  & {\bf BiNoM}\\
{\bf Authors}                              & {B\"uchel \emph{et al.}}          & {Ruebenacker \emph{et al.}}   & {Zinovyev \emph{et al.}} \\
%{\bf Tested with} & PID-BioCarta-ceramide signaling (L2)
%Reacomte-Homarus americanus (L3) & Reactome-Homarus americanus (Level 3) & ceramide signaling von PID-BioCarta \\
{\bf Version}                       & 1.0                               & 1.0 (Build 119)               & 2.0  \\
{\bf Release date}                  & 2012-04-02                        & 2010-02-11                    & 2012-04-12  \\
\midrule
{\bf BioPAX Input Level}            & Levels 2 and 3                    & Levels 2 and 3                & Level 3 \\
{\bf SBML  Output Level}            & Level 3 Version 1                 & Level 2 Version 4             & Level 2 Version 4, Beta \\
                                    &                                   &                               & Version for Level 3 \\
\midrule
{\bf Graphical user interface}      & \checkmark                        & \checkmark                    & \checkmark \\
{\bf Qualitative model support}     & \checkmark                        & -                             & -  \\
{\bf Valid SBML}                    & \checkmark                        & -                             & -  \\
{\bf Complete}                      & \checkmark                        & -                             & \checkmark\\
{\bf No duplicate entities}         & \checkmark                        & \checkmark                    & \checkmark \\
{\bf Robustness}                    & \checkmark                        & $\circ$                       & $\circ$\\
\midrule
{\bf Compartments}                  & \checkmark                        & \checkmark                    & \checkmark \\
{\bf Stoichiometry}                 & \checkmark                        & -                             & - \\
{\bf SBO terms}                     & \checkmark                        & -                             & - \\
{\bf Xrefs converted into CV terms} & \checkmark                        & -                             & - \\
{\bf Augment model}                 & \checkmark                        & -                             & - \\
{\bf Provenance}                    & \checkmark                        & -                             & - \\
\midrule
\end{tabular}
}
{
This table compares three applications which are able to translate BioPAX into SBML.
The checkmark (\checkmark) indicates that the criterion is completely fulfilled, the circle ($\circ$) shows that the criterion is partially fulfilled, and the minus (-) is used if it is not fulfilled.
A conversion is \emph{valid} if the validator from \url{sbml.org} reports no errors in the converted SBML file and the converted model is \emph{complete} if no BioPAX entity is missing.
The \emph{No duplicate entities} criterion is important for modeling purposes to guarantee that a species is only mentioned once.
A converter is \emph{robust} if it can handle all tested files from the Pathway Interaction Database and is able to convert a BioPAX file, which contains no \Pathway{} element.
The \emph{Compartments}, \emph{Stoichiometry}, and \emph{Uses SBO terms} criteria denote if this information is translated into SBML and if the SBML species are denoted with the corresponding SBO term.
Additionally, it is checked if the BioPAX cross-references (\texttt{Xrefs}) are translated into SBML controlled vocabulary terms (CV terms) and if the SBML model is \emph{augmented} with further information, such as Entrez Gene IDs.
Finally, the \emph{provenance} criterion denotes if the file history and conversion tool information is saved in the converted SBML file.
}
\end{table*}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% END %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
The Pathway Interaction Database (PID) is a curated and peer-reviewed pathway database containing human pathways with molecular signaling and regulatory events provided by the Nature Cancer Institute, BioCarta, and Reactome.
All pathways are provided in XML, BioPAX Level~2, and Level~3 format.

The BioPAX format is perfectly suitable to encode pathway relations and reactions that can be further used for visualization or pathway analysis. However, this format also has its limitations.
Many applications, especially for simulation and modeling of biological networks, use the SBML format \citep{Zinovyev2008, Funahashi2007, Hoops2006}.
Therefore, a few importers and converters for BioPAX into SBML have been developed.
BioPAX \Entity{} elements, which can be genes, proteins, small molecules, etc., can be translated into SBML \species{} and the type of the BioPAX \Entity{} can be encoded as SBO term or MIRIAM annotation of the species itself.
Relations between \Entity{} elements, corresponding to edges in a pathway graph, are also provided with detailed information in BioPAX.
These relations can be transports, biochemical reactions, complex assemblies, etc.
At this point, most translations to SBML usually produce errors or have a massive loss of information because the SBML core specification only provides reactions, which represents real biochemical reactions with substrates, products and enzymes.
Processes, such as modulation of an entity by another one, \NEW{cannot} directly be encoded as a reaction, at least not without knowing the exact chemical equation.
Hence, former conversion approaches from BioPAX to SBML did either incorrectly convert those relations to reactions or simply removed them during the translation.
To fill this gap, the SBML community has recently developed the \qual{} specification, which allows users to model arbitrary transitions between species.

Furthermore, the models themselves just provide the base for further analysis or visualization methods.
Other applications, such as Clandestine or COPASI, focus on visualization, simulation, analysis, etc. of those models.
Therefore, most of those applications have certain requirements on the models.
For example, to uniquely map mass spectrometry data on a model, it may be required for the model to have UniProt IDs.
To match mRNA expression data or perform gene set enrichment analyses, Entrez Gene identifiers might be required.
Consequently, we provide all annotations that we could gather from the input BioPAX files also in the SBML files and further annotate all \species{} with a plethora of additional identifiers.

The \qual{} extension has been created recently and, thus, might not be supported by all applications, yet.
Therefore, we decided to build joint SBML core and \qual{} models.
All our SBML files contain a \model{} that corresponds to the SBML core specification, and an additional \qualitativeModel{} that contains all relations.
These models are compatible with older applications, that do not yet support \qual{} but still can read all \species{} and \reactions.
Newer applications that are ready to handle relations can read the additional \qual{} model and process all information that was also available in the BioPAX file.

We converted both, the Level~2 and the Level~3 BioPAX files to SBML core, including the \qual{} extension.
The reason for converting both levels was the additional description possibility of gene-regulatory networks and genetic interactions in BioPAX Level~3, which is not supported by Level~2 pathway models.
Since older simulation applications still work with BioPAX Level~2, we also translated these files into SBML in order to prevent loss of information and to be able to use these models, too.
All models are available at \url{http://www.cogsys.cs.uni-tuebingen.de/downloads/Qualitative-Models/}.
Furthermore, we provide our webtool BioPAX2SBML for further BioPAX translations at \url{http://webservices.cs.uni-tuebingen.de/}.



\subsection{Comparison to other BioPAX to SBML converters}
Only a few approaches exist to convert BioPAX to SBML and the existing ones use a simple one-to-one conversion without augmenting the file content for further modeling steps.
This might be due to the fact that ``the inter-conversion between BioPAX and SBML is not trivial as both formats were developed for different purposes" \citep{Bauer-Mehren2009}.
Sybill \citep{Ruebenacker2009} and BiNoM \citep{Zinovyev2008} are two approaches that can be used for the translation of BioPAX into SBML but none of them is able to appropriately translate signaling networks.
Sybill is a stand-alone tool that is also integrated in the quantitative modeling environment VCell \citep{Slepchenko2003}.
In contrast, BiNoM is a Cytoscape plugin, which offers the possibility to open BioPAX Level~3 files.
It mainly focusses on the visualization of BioPAX files and not on the actual conversion into SBML.
Table~\ref{tab:Comparison} compares these programs based on defined criteria.

Sybill converts BioPAX Level~2 and Level~3 files and has a very comfortable graphical user interface allowing the user to manipulate the conversion result.
Unfortunately, the converted SBML files are not complete and the validator from \url{sbml.org} reports errors, because species involved in several reactions are missing in the \texttt{listOfSpecies}.
Additionally, some groups and pathway links are missing, too.
BiNoM generates a complete conversion result, but the validator also reports errors due to empty \texttt{listOf} elements and due to the wrong order of these elements.
In contrast to Sybill, BiNoM converts BioPAX files without a pathway element, but is only able to handle a small number of BioPAX files from the Pathway Interaction Database.
Another feature of BiNoM is that it can separately visualize reaction networks, pathway structure and protein-protein interaction networks out of one BioPAX file.

All approaches avoid the translation of duplicate species.
Only BioPAX2SBML converter uses the \qual{} extension and SBO terms for detailed species description, translates the BioPAX cross-references (\texttt{Xrefs}) into SBML CV terms, and augments the \NEW{SBML} file content with further database cross-references.



\section{Conclusion}
Conversion between different formats is important in all parts of computer science.
In many cases, conversion leads to errors or a loss of information.
The BioPAX to SBML conversion is such an example.
Due to limitations of the SBML core specification, it was not possible to include all relationships between reactive species from BioPAX files in SBML files, while producing correct SBML code.
But with SBML Level~3 Version 1 and the addition of extensions to the specifications, in particular the Qualitative Models extension (\qual), it is now possible to create accurate and specification-conform SBML code.
Using this extension, we produced error-free SBML models while minimizing or even eliminating the loss of information during the translation.

The SBML models, provided along with this publication, consist of SBML \species{} and, wherever possible, exact \reaction{} equations.
All relations from the BioPAX documents that could not be converted to exact reactions have been included as qualitative transitions between qualitative species.
Additional information, such as various identifiers or the type of an entity, are encoded as SBO terms or MIRIAM URNs of the corresponding elements.
By utilizing the KEGG API it was even possible to complement the translated BioPAX documents with a wealth of information from further databases, such as Entrez Gene, KEGG, etc.

Compared to existing conversion approaches with similar scope, BioPAX2SBML conversions result in comprehensive and correct SBML models, created for all pathways in the Nature Pathway Interaction Database.
These models can easily be used, e.g., for further simulation and modeling steps, without having to deal with incorrect input file formats or error-prone conversions.

\section*{Acknowledgements}
We wish to acknowledge the Qualitative Models and JSBML teams.

\paragraph{Funding\textcolon}
Federal Ministry of Education and Research (BMBF, Germany) in the National
Genome Research Network (NGFN+) under grant number 01GS08134 and Virtual Liver
Network under grant number 0315756.
\paragraph{Conflict of interest\textcolon} None declared.


\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
%\bibliographystyle{plain}
%
%\bibliography{document}

\begin{thebibliography}{}

\bibitem[Bauer-Mehren {\em et~al.}(2009)Bauer-Mehren, Furlong, and
  Sanz]{Bauer-Mehren2009}
Bauer-Mehren, A., Furlong, L.~I., and Sanz, F. (2009).
\newblock Pathway databases and tools for their exploitation: benefits, current
  limitations and challenges.
\newblock {\em Mol Syst Biol\/}, {\bf 5}, 290.

\bibitem[Berenguier {\em et~al.}(2011)Berenguier, Chaouiya, Naldi, Thieffry,
  and van Iersel]{QualSpecification}
Berenguier, D., Chaouiya, C., Naldi, A., Thieffry, D., and van Iersel, M.~P.
  (2011).
\newblock {Qualitative Models} (qual).
\newblock Specification available at
  \url{http://sbml.org/Community/Wiki/SBML_Level_3_Proposals/Qualitative_Models}.
  Accessed 2012 Mar 22.

\bibitem[Courtot {\em et~al.}(2011)Courtot, Juty, Kn\"{u}pfer, Waltemath,
  Zhukova, Dr\"{a}ger, Dumontier, Finney, Golebiewski, Hastings, Hoops,
  Keating, Kell, Kerrien, Lawson, Lister, Lu, Machne, Mendes, Pocock,
  Rodriguez, Villeger, Wilkinson, Wimalaratne, Laibe, Hucka, and
  Nov\`{e}re]{SBO}
Courtot, M., Juty, N., Kn\"{u}pfer, C., Waltemath, D., Zhukova, A., Dr\"{a}ger,
  A., Dumontier, M., Finney, A., Golebiewski, M., Hastings, J., Hoops, S.,
  Keating, S., Kell, D.~B., Kerrien, S., Lawson, J., Lister, A., Lu, J.,
  Machne, R., Mendes, P., Pocock, M., Rodriguez, N., Villeger, A., Wilkinson,
  D.~J., Wimalaratne, S., Laibe, C., Hucka, M., and Nov\`{e}re, N.~L. (2011).
\newblock Controlled vocabularies and semantics in systems biology.
\newblock {\em Mol Syst Biol\/}, {\bf 7}, 543.

\bibitem[Demir {\em et~al.}(2010)Demir, Cary, Paley, Fukuda, Lemer, Vastrik,
  Wu, D'Eustachio, Schaefer, Luciano, Schacherer, Martinez-Flores, Hu,
  Jimenez-Jacinto, Joshi-Tope, Kandasamy, Lopez-Fuentes, Mi, Pichler,
  Rodchenkov, Splendiani, Tkachev, Zucker, Gopinath, Rajasimha, Ramakrishnan,
  Shah, Syed, Anwar, Babur, Blinov, Brauner, Corwin, Donaldson, Gibbons,
  Goldberg, Hornbeck, Luna, Murray-Rust, Neumann, Reubenacker, Samwald, van
  Iersel, Wimalaratne, Allen, Braun, Whirl-Carrillo, Cheung, Dahlquist, Finney,
  Gillespie, Glass, Gong, Haw, Honig, Hubaut, Kane, Krupa, Kutmon, Leonard,
  Marks, Merberg, Petri, Pico, Ravenscroft, Ren, Shah, Sunshine, Tang, Whaley,
  Letovksy, Buetow, Rzhetsky, Schachter, Sobral, Dogrusoz, McWeeney, Aladjem,
  Birney, Collado-Vides, Goto, Hucka, Nov\`{e}re, Maltsev, Pandey, Thomas,
  Wingender, Karp, Sander, and Bader]{Demir2010}
Demir, E., Cary, M.~P., Paley, S., Fukuda, K., Lemer, C., Vastrik, I., Wu, G.,
  D'Eustachio, P., Schaefer, C., Luciano, J. {\em et~al.} (2010).
\newblock The {BioPAX} community standard for pathway data sharing.
\newblock {\em Nat Biotechnol\/}, {\bf 28}(9), 935--942.

\bibitem[Dr\"ager {\em et~al.}(2011)Dr\"ager, Rodriguez, Dumousseau, D\"orr,
  Wrzodek, LeNov\`{e}re, Zell, and Hucka]{Draeger2011}
Dr\"ager, A., Rodriguez, N., Dumousseau, M., D\"orr, A., Wrzodek, C.,
  LeNov\`{e}re, N., Zell, A., and Hucka, M. (2011).
\newblock {JSBML: a flexible Java library for working with SBML.}
\newblock {\em Bioinformatics\/}, {\bf 27}(15), 2167--2168.

\bibitem[{European Bioinformatics Institute -- Computational Systems
  Neurobiology Group}(2011){European Bioinformatics Institute -- Computational
  Systems Neurobiology Group}]{SBFC}
{European Bioinformatics Institute -- Computational Systems Neurobiology Group}
  (2011).
\newblock {System Biology Format Converter (SBFC)}.
\newblock Software available from
  \url{http://www.ebi.ac.uk/compneur-srv/sbml/converters/SBMLtoBioPax.html}
  Accessed 2012 Mar 22.

\bibitem[Funahashi {\em et~al.}(2007)Funahashi, Jouraku, Matsuoka, and
  Kitano]{Funahashi2007}
Funahashi, A., Jouraku, A., Matsuoka, Y., and Kitano, H. (2007).
\newblock {Integration of CellDesigner and SABIO-RK.}
\newblock {\em In Silico Biol\/}, {\bf 7}(2 Suppl), S81--S90.

\bibitem[Hoops {\em et~al.}(2006)Hoops, Sahle, Gauges, Lee, Pahle, Simus,
  Singhal, Xu, Mendes, and Kummer]{Hoops2006}
Hoops, S., Sahle, S., Gauges, R., Lee, C., Pahle, J., Simus, N., Singhal, M.,
  Xu, L., Mendes, P., and Kummer, U. (2006).
\newblock Copasi--a complex pathway simulator.
\newblock {\em Bioinformatics\/}, {\bf 22}(24), 3067--3074.

\bibitem[Hucka {\em et~al.}(2003)Hucka, Finney, Sauro, Bolouri, Doyle, Kitano,
  Arkin, Bornstein, Bray, Cornish-Bowden, Cuellar, Dronov, Gilles, Ginkel, Gor,
  Goryanin, Hedley, Hodgman, Hofmeyr, Hunter, Juty, Kasberger, Kremling,
  Kummer, Nov\`{e}re, Loew, Lucio, Mendes, Minch, Mjolsness, Nakayama, Nelson,
  Nielsen, Sakurada, Schaff, Shapiro, Shimizu, Spence, Stelling, Takahashi,
  Tomita, Wagner, Wang, and Forum]{Hucka2003}
Hucka, M., Finney, A., Sauro, H.~M., Bolouri, H., Doyle, J.~C., Kitano, H.,
  Arkin, A.~P., Bornstein, B.~J., Bray, D., Cornish-Bowden, A., Cuellar, A.~A.,
  Dronov, S., Gilles, E.~D., Ginkel, M., Gor, V., Goryanin, I.~I., Hedley,
  W.~J., Hodgman, T.~C., Hofmeyr, J.-H., Hunter, P.~J., Juty, N.~S., Kasberger,
  J.~L., Kremling, A., Kummer, U., Nov\`{e}re, N.~L., Loew, L.~M., Lucio, D.,
  Mendes, P., Minch, E., Mjolsness, E.~D., Nakayama, Y., Nelson, M.~R.,
  Nielsen, P.~F., Sakurada, T., Schaff, J.~C., Shapiro, B.~E., Shimizu, T.~S.,
  Spence, H.~D., Stelling, J., Takahashi, K., Tomita, M., Wagner, J., Wang, J.,
  and Forum, S. B. M.~L. (2003).
\newblock {The systems biology markup language (SBML): a medium for
  representation and exchange of biochemical network models.}
\newblock {\em Bioinformatics\/}, {\bf 19}(4), 524--531.

\bibitem[Kanehisa {\em et~al.}(2006)Kanehisa, Goto, Hattori, Aoki-Kinoshita,
  Itoh, Kawashima, Katayama, Araki, and Hirakawa]{Kanehisa2006}
Kanehisa, M., Goto, S., Hattori, M., Aoki-Kinoshita, K.~F., Itoh, M.,
  Kawashima, S., Katayama, T., Araki, M., and Hirakawa, M. (2006).
\newblock From genomics to chemical genomics: new developments in {KEGG}.
\newblock {\em Nucleic Acids Res\/}, {\bf 34}(Database issue), D354--D357.

\bibitem[Mi {\em et~al.}(2011)Mi, Muruganujan, Demir, Matsuoka, Funahashi,
  Kitano, and Thomas]{Mi2011}
Mi, H., Muruganujan, A., Demir, E., Matsuoka, Y., Funahashi, A., Kitano, H.,
  and Thomas, P.~D. (2011).
\newblock {BioPAX support in CellDesigner.}
\newblock {\em Bioinformatics\/}, {\bf 27}(24), 3437--3438.

\bibitem[Nov\`{e}re {\em et~al.}(2005)Nov\`{e}re, Finney, Hucka, Bhalla,
  Campagne, Collado-Vides, Crampin, Halstead, Klipp, Mendes, Nielsen, Sauro,
  Shapiro, Snoep, Spence, and Wanner]{Novere2005}
Nov\`{e}re, N.~L., Finney, A., Hucka, M., Bhalla, U.~S., Campagne, F.,
  Collado-Vides, J., Crampin, E.~J., Halstead, M., Klipp, E., Mendes, P.,
  Nielsen, P., Sauro, H., Shapiro, B., Snoep, J.~L., Spence, H.~D., and Wanner,
  B.~L. (2005).
\newblock Minimum information requested in the annotation of biochemical models
  {(MIRIAM)}.
\newblock {\em Nat Biotechnol\/}, {\bf 23}(12), 1509--1515.

\bibitem[R\"ubenacker {\em et~al.}(2009)R\"ubenacker, Moraru, Schaff, and
  Blinov]{Ruebenacker2009}
R\"ubenacker, O., Moraru, I.~I., Schaff, J.~C., and Blinov, M.~L. (2009).
\newblock {Integrating BioPAX pathway knowledge with SBML models.}
\newblock {\em IET Syst Biol\/}, {\bf 3}(5), 317--328.

\bibitem[Schaefer {\em et~al.}(2009)Schaefer, Anthony, Krupa, Buchoff, Day,
  Hannay, and Buetow]{Schaefer2009}
Schaefer, C.~F., Anthony, K., Krupa, S., Buchoff, J., Day, M., Hannay, T., and
  Buetow, K.~H. (2009).
\newblock {PID: the Pathway Interaction Database.}
\newblock {\em Nucleic Acids Res\/}, {\bf 37}(Database issue), D674--D679.

\bibitem[Slepchenko {\em et~al.}(2003)Slepchenko, Schaff, Macara, and
  Loew]{Slepchenko2003}
Slepchenko, B.~M., Schaff, J.~C., Macara, I., and Loew, L.~M. (2003).
\newblock Quantitative cell biology with the virtual cell.
\newblock {\em Trends Cell Biol\/}, {\bf 13}(11), 570--576.

\bibitem[Smoot {\em et~al.}(2011)Smoot, Ono, Ruscheinski, Wang, and
  Ideker]{Smoot2011a}
Smoot, M.~E., Ono, K., Ruscheinski, J., Wang, P.-L., and Ideker, T. (2011).
\newblock Cytoscape 2.8: new features for data integration and network
  visualization.
\newblock {\em Bioinformatics\/}, {\bf 27}(3), 431--432.

\bibitem[Zinovyev {\em et~al.}(2008)Zinovyev, Viara, Calzone, and
  Barillot]{Zinovyev2008}
Zinovyev, A., Viara, E., Calzone, L., and Barillot, E. (2008).
\newblock {BiNoM: a Cytoscape plugin for manipulating and analyzing biological
  networks.}
\newblock {\em Bioinformatics\/}, {\bf 24}(6), 876--877.

\end{thebibliography}

\include{Buechel_Appendix}

\end{document}
